be/src/exprs/expr.cpp

```diff
@@ -654,6 +689,83 @@ ColumnRef* Expr::get_column_ref() {
     return nullptr;
 }
 
+StatusOr<LLVMDatum> Expr::generate_ir(ExprContext* context, const llvm::Module& module, llvm::IRBuilder<>& b,
+                                      const std::vector<LLVMDatum>& datums) const {
+    if (!is_compilable()) {
+        return Status::NotSupported("JIT expr not supported");
+    }
+
+    ASSIGN_OR_RETURN(auto datum, generate_ir_impl(context, module, b, datums))
+    // Unoin null.
+    if (this->is_nullable()) {
+        // TODO(Yueyang): Check this.
+        for (auto& input : datums) {
+            datum.null_flag = b.CreateOr(datum.null_flag, input.null_flag);
+        }
+    }
+    return datum;
+}
+
+void Expr::get_uncompilable_exprs(std::vector<Expr*>& exprs) {
+    if (!this->is_compilable()) {
+        exprs.emplace_back(this);
+        return;
+    }
+    for (auto child : this->children()) {
+        child->get_uncompilable_exprs(exprs);
+    }
+}
+
+void Expr::get_jit_exprs(std::vector<Expr*>& exprs) {
+    if (!this->is_compilable()) {
+        exprs.emplace_back(this);
+        return;
+    }
+    for (auto child : this->children()) {
+        child->get_jit_exprs(exprs);
+    }
+    exprs.emplace_back(this);
+}
+
+// This method attempts to traverse the entire expression tree from the current expression downwards, seeking to replace expressions with JITExprs.
+// This method searches from top to bottom for compilable expressions.
+// Once a compilable expression is found, it skips over its compilable subexpressions and continues the search downwards.
+Status Expr::replace_compilable_exprs(Expr** expr, ObjectPool* pool) {
+    std::vector<Expr*> next_exprs;
+    if ((*expr)->should_compile()) {
+        // If the current expression is compilable, we will replace it with a JITExpr.
+        // This expression and its compilable subexpressions will be compiled into a single function.
+        *expr = JITExpr::create(pool, *expr);
+        // Skip over the compilable subexpressions.
+        (*expr)->get_uncompilable_exprs(next_exprs);
+    } else {
+        // Search for compilable expressions in the subexpressions.
+        next_exprs = _children;
+    }
+
+    for (auto& child : next_exprs) {
+        RETURN_IF_ERROR(child->replace_compilable_exprs(&child, pool));
+    }
+    return Status::OK();
+}
+
+bool Expr::should_compile() const {
+    if (!is_compilable() || _children.empty()) {
+        return false;
+    }
+
+    for (auto child : _children) {
+        // If an expr is compilable, and it has compilable child nodes that are not leaf nodes,
+        // compiling these compilable nodes into one node via JIT will provide benefits.
+        // The 'literal' is special. It is compilable, but it doesn't have any child nodes
+        if (child->is_compilable() && !child->children().empty()) {
+            return true;
+        }
```

Comment:
`col a + col b`, the `+` is not compile?

Reply:
Yes, if there's only a single expression like ﻿a + b, compiling it wouldn't yield any benefits.